sycl/plugins/opencl/CMakeLists.txt

```diff
@@ -0,0 +1,57 @@
+#To-Do:
+#1. Figure out why CMP0057 has to be set. Should have been taken care of earlier in the build
+#2. Use AddLLVM to modify the build and access config options
+#cmake_policy(SET CMP0057 NEW)
+#include(AddLLVM)
+
+# Plugin for OpenCL
+# Create Shared library for libpi_opencl.so.
+#To-DO: remove dependency on pi.hpp in sycl project.
+add_library(pi_opencl SHARED
+  "${sycl_inc_dir}/CL/sycl/detail/pi.hpp"
+  "pi_opencl.cpp"
+  )
+
+add_dependencies(pi_opencl
+  ocl-icd
+  ocl-headers
+)
+
+set_target_properties(pi_opencl PROPERTIES LINKER_LANGUAGE CXX)
+
+#preprocessor definitions for compiling a target's sources. We do not need it for pi_opencl
+target_include_directories(pi_opencl PRIVATE "${sycl_inc_dir}")
+
+#link pi_opencl with OpenCL headers and ICD Loader.
+target_link_libraries( pi_opencl
+    PRIVATE OpenCL::Headers
+    PRIVATE ${OpenCL_LIBRARIES}
+)
+
+if (SYCL_USE_LIBCXX)
```

Comment:
Is there a better way to apply common workarounds for building a new C++ component?

Reply:
@ilyastepykin This section of the CMakeFile mirrors the changes from the CMakefile in the sources directory. These changes were added by you. Can you comment why these changes were added ? Do they address a specific requirement?

Reply:
These changes were added to allow libc++ library to be used instead of libstdc++ for compiling host and device code. For that case SYCL runtime library has to be build and linked with libc++ as well. So this code below does that if SYCL_USE_LIBCXX option is set. Looks like the same thing has to be done for plugin as well.

Reply:
It might be worth making these settings `PUBLIC` in `sycl/source/CMakeLists.txt`, i.e.:
```cmake
if (SYCL_USE_LIBCXX)
    if ((CMAKE_CXX_COMPILER_ID STREQUAL "GNU") OR
        (CMAKE_CXX_COMPILER_ID STREQUAL "Clang"))
        target_compile_options(sycl PUBLIC -nostdinc++)
        if ((NOT (DEFINED SYCL_LIBCXX_INCLUDE_PATH)) OR (NOT (DEFINED SYCL_LIBCXX_LIBRARY_PATH)))
            message(FATAL_ERROR "When building with libc++ SYCL_LIBCXX_INCLUDE_PATHS and"
                                "SYCL_LIBCXX_LIBRARY_PATH should be set")
        endif()
        target_include_directories(sycl PUBLIC "${SYCL_LIBCXX_INCLUDE_PATH}")
        target_link_libraries(sycl PUBLIC "-L${SYCL_LIBCXX_LIBRARY_PATH}" -nodefaultlibs -lc++ -lc++abi -lm -lc -lgcc_s -lgcc)
    else()
        message(FATAL_ERROR "Build with libc++ is not yet supported for this compiler")
    endif()
else()
```


If we make the plugins link with the `sycl` target they will then "inherit" the public settings.
There might be unintended consequences though, so this would need experimentation.

Reply:
The following PR seems to provide such a helper function or what could be the base of a helper function: https://github.com/intel/llvm/pull/685/files#diff-8b14e80fb16152854e3f34abccd44a97R7